hydrosdk/predictor.py

```diff
@@ -1,72 +1,165 @@
+
+        try:
+            response = self.impl.send_data(model_spec=self.model_spec, inputs=inputs_as_proto)
+
+            if self.return_type == PredictorDT.DF:
+                return self.predict_resp_to_df(response=response)
+
+            elif self.return_type == PredictorDT.DICT_NP_ARRAY:
+                return self.predict_resp_to_dict_nparray(response=response)
+
+            elif self.return_type == PredictorDT.DICT_PYTHON:
+                return self.predict_resp_to_dict_pydtype(response=response)
+        except Exception as err:
+            raise Exception(f"Failed to predict.{str(err)} ")
+
+    @staticmethod
+    def predict_resp_to_dict_pydtype(response: PredictResponse) -> dict:
+        output_tensors_dict = {}
+        for tensor_name, tensor_proto in response.outputs.items():
+            dims = [dim.size for dim in tensor_proto.tensor_shape.dim]
+            value = getattr(tensor_proto, DTYPE_TO_FIELDNAME[tensor_proto.dtype])
+
+            # If no dims specified in TensorShapeProto, then it is scalar
+            if dims:
+                value = np.reshape(value, dims).tolist()
+                output_tensors_dict[tensor_name] = value
+            else:
+                output_tensors_dict[tensor_name] = value[0]
+
+        return output_tensors_dict
+
+    @staticmethod
+    def predict_resp_to_dict_nparray(response: PredictResponse) -> dict:
+        output_tensors_dict = {}
+        for tensor_name, tensor_proto in response.outputs.items():
+            array_shape = [dim.size for dim in tensor_proto.tensor_shape.dim]
+            np_dtype = proto2np_dtype(tensor_proto.dtype)
+            value = getattr(tensor_proto, DTYPE_TO_FIELDNAME[tensor_proto.dtype])
+            np_array_value = np.array(value, dtype=np_dtype)
+
+            # If no dims specified in TensorShapeProto, then it is scalar
+            if array_shape:
+                output_tensors_dict[tensor_name] = np_array_value.reshape(*array_shape)
+            else:
+                output_tensors_dict[tensor_name] = np.asscalar(np_array_value)
+
+        return output_tensors_dict
+
+    @staticmethod
+    def predict_resp_to_df(response: PredictResponse) -> pd.DataFrame:
+        response_dict: Dict[str, np.array] = PredictServiceClient.predict_resp_to_dict_nparray(response)
+        return pd.DataFrame(response_dict)
+
+
+class Predictable(ABC):
+    @abstractmethod
+    def predictor(self):
+        pass
+
+
+class PredictableApplication(Predictable):
+    """Adds Predictor functionality"""
+
+    def predictor(self, ssl=False, return_type=PredictorDT.DICT_NP_ARRAY) -> PredictServiceClient:
+        """
+
+        :param ssl:
+        :param return_type:
+        :return:
+        """
+        if ssl:
+            self.channel = self.cluster.grpc_secure()
+        else:
+            self.channel = self.cluster.grpc_insecure()
+
+        self.impl = MetricableImplementation(self.channel)
+        self.predictor_return_type = return_type
```

Comment:
❌ Both cluster.grpc_secure() and cluster.grpc_insecure() methods accept options, compression settings and security parameters.

Moreover, I think we must reconsider passing the grpc channel options here — in my opinion, if user already sets grpc address in Cluster ctor, could it be simpler to move the channel parameters there?
After all, the user will get simple Cluster.grpc_channel() method without parameters, which will hide ugly details of Cluster networking.

Reply:
Like this?